### PR TITLE
Make use of smart pointers and references

### LINKED ISSUE #314: Choosing reference over pointer
It is recommended to use references when we can, and pointers when we have to.
(Reference: https://isocpp.org/wiki/faq/references#refs-vs-ptrs)

We can replace `Device *_parent` to `Device & _parent` as I see that, its possible. We initialize while activating plugin and never modify it. Also, we have a method (below) that actually gives, **a reference to a device**, not pointer. 
```cpp
Device &device(uint64_t uuid);
```

Consider, below piece of application code:
```cpp
Device &device = dc.device(); // we get reference to device
auto action = std::make_shared<Action>(&device); // why pass its address ?

// Why not pass it as-is?
auto action = std::make_shared<Action>(device);
```
I just noticed this.
@julianoes, was it done intentionally for some vital reason? If not, then we can find similar usages and replace them properly.

### PR DESCRIPTION
**Pull request by shakthi-prashanth-m**
This commit replaces raw pointers by smart pointers and references as applicable.
We have to use references when we can, and pointers when we have to.

Smart pointers provide automatic resource management. In other words, they implement **Resource Acquisition Is Initialization** programming idiom. The main goal of this idiom is to ensure that resource acquisition occurs at the same time that the object is initialized, so that all resources for the object are created and made ready in one line of code.

 `std::unique_ptr` is a smart pointer that owns and manages another object through a pointer and disposes of that object when the unique_ptr goes out of scope.

`std::shared_ptr` is a smart pointer that retains shared ownership of an object through a pointer. Several shared_ptr objects may own the same object. The object is destroyed and its memory deallocated when either of the following happens:
- the last remaining `shared_ptr` owning the object is destroyed;
- the last remaining `shared_ptr` owning the object is assigned another pointer via `operator=` or `reset()`. 

    
**References:**
- https://isocpp.org/wiki/faq/references#refs-vs-ptrs
- https://docs.microsoft.com/en-us/cpp/cpp/smart-pointers-modern-cpp
- http://en.cppreference.com/w/cpp/memory/unique_ptr
- http://en.cppreference.com/w/cpp/memory/shared_ptr

Fixes #314

### CODE REVIEW

**Comment by shakthi-prashanth-m** [comment]
I think, we should use `make_unique` construct to make them smart pointer. I shall change it.

**Comment by shakthi-prashanth-m** [comment]
`make_unique` constructs an object of type T and wraps it in a std::unique_ptr. Its available in C++14. Its fine to use raw pointer and assign to a `unique_ptr` in C+11.

File: core/dronecore_impl.cpp [hunk 1]
```diff
@@ -330,7 +330,7 @@ void DroneCoreImpl::create_device_if_not_existing(const uint8_t system_id)
     }
 
     // Create both lists in parallel
```
Comment (shakthi-prashanth-m) [inline core/dronecore_impl.cpp]: Actually, this comment makes no sense now :) I shall remove it in next commit.

**Comment by shakthi-prashanth-m** [comment]
Executed `SitlTest.TwoConnections` and `SitlTest.AsyncConnect`.
```
[==========] Running 2 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 2 tests from SitlTest
[ RUN      ] SitlTest.TwoConnections
Not autostarting SITL (use AUTOSTART_SITL=1)
started
[04:37:56|Info ] New device on: 127.0.0.1:14556 (udp_connection.cpp:207)
[04:37:56|Error] Remote IP unknown (udp_connection.cpp:136)
[04:37:56|Error] send fail (dronecore_impl.cpp:90)
[04:37:56|Info ] New device on: 127.0.0.1:14557 (udp_connection.cpp:207)
[04:37:57|Debug] Discovered 4294967298 (dronecore_impl.cpp:320)
[04:37:57|Debug] MAVLink: info: DISARMED by auto disarm on land (device.cpp:251)
[04:37:57|Debug] MAVLink: info: [logger] file: rootfs/fs/microsd/log/2018-03-13/1 (device.cpp:251)
[04:37:57|Debug] MAVLink: info: data link #0 lost (device.cpp:251)
[04:37:57|Debug] MAVLink: critical: First waypoint too far away: 7094 meters, 900 max (device.cpp:251)
[04:37:57|Debug] MAVLink: info: ARMED by arm/disarm component command (device.cpp:251)
[04:37:57|Debug] MAVLink: info: Using minimum takeoff altitude: 2.50 m (device.cpp:251)
[04:37:57|Debug] MAVLink: info: Takeoff detected (device.cpp:251)
[04:37:57|Debug] MAVLink: critical: Planned mission fails check. Please upload again. (device.cpp:251)
[04:37:57|Debug] MAVLink: critical: Planned mission fails check. Please upload again. (device.cpp:251)
[04:37:57|Debug] MAVLink: critical: Planned mission fails check. Please upload again. (device.cpp:251)
[04:37:57|Debug] MAVLink: critical: Planned mission fails check. Please upload again. (device.cpp:251)
[04:37:57|Debug] MAVLink: info: data link #0 regained (device.cpp:251)
[04:37:57|Debug] MAVLink: info: data link #0 regained (device.cpp:251)
found device with UUID: 4294967298
deleting it
[04:38:01|Info ] Clearing devices. (dronecore_impl.cpp:31)
[04:38:01|Info ] Destructor called. (device.cpp:43)
[04:38:01|Info ] Clearing connections. (dronecore_impl.cpp:37)
[04:38:01|Info ] UDP destruct... (udp_connection.cpp:35)
[04:38:01|Info ] Conn destruct... (connection.cpp:14)
[04:38:01|Info ] UDP destruct... (udp_connection.cpp:35)
[04:38:01|Info ] Conn destruct... (connection.cpp:14)
[04:38:02|Info ] New device on: 127.0.0.1:14557 (udp_connection.cpp:207)
[04:38:03|Debug] Discovered 4294967298 (dronecore_impl.cpp:320)
created and started yet again
[04:38:06|Info ] Clearing devices. (dronecore_impl.cpp:31)
[04:38:06|Info ] Destructor called. (device.cpp:43)
[04:38:06|Info ] Clearing connections. (dronecore_impl.cpp:37)
[04:38:06|Info ] UDP destruct... (udp_connection.cpp:35)
[04:38:06|Info ] Conn destruct... (connection.cpp:14)
exiting
[       OK ] SitlTest.TwoConnections (14016 ms)
[ RUN      ] SitlTest.AsyncConnect
Not autostarting SITL (use AUTOSTART_SITL=1)
waiting for device to appear...
[04:38:10|Info ] New device on: 127.0.0.1:14557 (udp_connection.cpp:207)
[04:38:10|Debug] MAVLink: critical: Planned mission fails check. Please upload again. (device.cpp:251)
[04:38:10|Debug] MAVLink: info: data link #1 lost (device.cpp:251)
[04:38:11|Debug] Discovered 4294967298 (dronecore_impl.cpp:320)
Found device with UUID: 4294967298
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
waiting for device to disappear...
[04:38:27|Info ] heartbeats timed out (device.cpp:256)
[04:38:27|Debug] Lost 4294967298 (dronecore_impl.cpp:328)
Lost device with UUID: 4294967298
[04:38:27|Info ] Clearing devices. (dronecore_impl.cpp:31)
[04:38:27|Info ] Destructor called. (device.cpp:43)
[04:38:27|Info ] Clearing connections. (dronecore_impl.cpp:37)
[04:38:27|Info ] UDP destruct... (udp_connection.cpp:35)
[04:38:27|Info ] Conn destruct... (connection.cpp:14)
[       OK ] SitlTest.AsyncConnect (21072 ms)
[----------] 2 tests from SitlTest (35088 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test case ran. (35088 ms total)
[  PASSED  ] 2 tests.
```

**Comment by julianoes** [comment]
@JonasVautherin when is a good time to merge this? :smile:

**Comment by JonasVautherin** [comment]
I'm finishing my review now, I will merge it in a few minutes. Sorry for the delay =/

**Pull request merged**